File: Projetos_C/Terceiro Semestre/Estrutura de Dados/Aula 11/Ex4.c
/*Usando pilha ou fila, determine se uma string de caracteres tem a seguinte forma:
xCx onde x é uma string composta por As e Bs.
Um exemplo de xCx é ABBBCABBB.
Em cada ponto você só poderá ler o próximo carater da string, para decidir se está ou não no padrão
esperado.*/

#include<stdio.h>
#include<stdlib.h>
#include<string.h>

#define STACKSIZE 100

struct stack {
	int top;
	char items[STACKSIZE];
};

int empty (struct stack *ps);
int pop (struct stack *ps);
void push (struct stack *ps, char x);

int main (void){
    char string[STACKSIZE];
    struct stack primeiraMetade;
    primeiraMetade.top=-1;
    int padrao=1;
    scanf("%s", string);
    int tamString=strlen(string);
    for(int i = tamString; i>=0; i--)
        push(&primeiraMetade, string[i]);
    for(int i=tamString/2+1; i<tamString; i++){
        if(string[i]!=pop(&primeiraMetade)){
            padrao=0;
            break;
        }
        else
            padrao=1;
    }

    if(padrao)
        printf("Esta no padrao");
    else
        printf("Esta fora do padrao");
    return 0;
}

int empty (struct stack *ps){
	if (ps->top==-1)
		return 1;
	return 0;	
}

int pop (struct stack *ps){
	int aux;
	if (empty(ps)){
		printf("ERRO- PILHA VAZIA");
		exit(1);
	}
	
	aux=ps->items[ps->top];
	ps->top--;
	return aux;
}

void push (struct stack *ps, char x){
	if (ps->top==STACKSIZE-1){
		printf("ERRO- PILHA CHEIA");
		exit(1);		
	}

	ps->top++;
	ps->items[ps->top]=x;
}

File: Projetos_Python/3 Semestre/Primeira Fase/Prova1/3-Pilha e Fila/Lista1/Ex4.py
STACKSIZE=100
class Pilha:

  def __init__(self):
    self.topo = -1
    self.valores = [None]*STACKSIZE

  def pilha_cheia(self):
    if self.topo == STACKSIZE - 1:
      return True
    return False

  def pilha_vazia(self):
    if self.topo == -1:
      return True
    return False

  def empilhar(self, valor):
    if self.pilha_cheia():
      print('A pilha está cheia')
      exit(1)
    self.topo += 1
    self.valores[self.topo] = valor

  def desempilhar(self):
    if self.pilha_vazia():
      print('A pilha está vazia')
      exit(1)
    aux=self.valores[self.topo]
    self.topo -= 1
    return aux

"""
4. Usando pilha ou fila, determine se uma string de caracteres tem a seguinte forma:
xCx onde x é uma string composta por As e Bs.
Um exemplo de xCx é ABBBCABBB.
Em cada ponto você só poderá ler o próximo carater da string, para decidir se está ou não no padrão
esperado.
"""

string = input()
posicao = string.find('C')
pilha = Pilha()
igual = False

for i in range(posicao - 1, -1, -1):
    pilha.empilhar(string[i])

for i in range(posicao + 1, len(string)):
    if not pilha.pilha_vazia() and string[i] == pilha.desempilhar():
        igual = True
    else:
        igual = False
        break

if igual:
    print("Igual")
else:
    print("Diferente")

  
  
